### PR TITLE
[rom] Add "pure" and "pre-hash" domain separators for SPHINCS+.

### PR DESCRIPTION
**Pull request by jadephilipoom**
Includes commits from https://github.com/lowRISC/opentitan/pull/23762; draft until that PR is merged.

Adds an OTP parameter to toggle between "pure" and "pre-hashed" SPHINCS+ mode, and code in the ROM to toggle between the two.

I did not add tests for the pre-hashed mode because, to my knowledge, test data is not available. @moidx @cfrantz are you aware of any existing pre-generated tests or have any thoughts on how to test this other than to just make some more test data a la `spx_verify_functest` to check that the opentitantool and device code agree?

Part of https://github.com/lowRISC/opentitan/issues/23144

### CODE REVIEW

File: hw/ip/otp_ctrl/data/BUILD
Comment (cfrantz) [inline hw/ip/otp_ctrl/data/BUILD:99]: I think we should consult with @moidx on this:  should the SPX+ variant be determined by OTP, or should we put something in the manifest to identify the signature variant?

**Comment by cfrantz** [comment]
LGTM, pending the resolution of how we decide between Pure and Pre-Hashed.

File: hw/ip/otp_ctrl/data/BUILD
Comment (moidx) [reply hw/ip/otp_ctrl/data/BUILD:99]: The specification recommends not using the same key in both pre-hashed and pure forms, so using this new OTP setting is ok. 

I prefer to use the existing `ROT_CREATOR_AUTH_CODESIGN_SPX_KEY_CONFIG0` so that we can make this decision at key provisioning time. 

https://github.com/lowRISC/opentitan/blob/6e698b4dfec2f9e61e091cbb9a866197681745f9/hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson#L430-L433

The key types can be augmented here: 

https://github.com/lowRISC/opentitan/blob/6e698b4dfec2f9e61e091cbb9a866197681745f9/sw/device/silicon_creator/lib/sigverify/spx_key.h#L56-L76

We can consider removing the config fields if we are not planning to use them, but take a look first. It is possible that the config field is part of the key struct used in spx sigverify.

Comment (cfrantz) [reply hw/ip/otp_ctrl/data/BUILD:99]: As I understand it, the guidance is that one should not use the same key for creating pure vs. pre-hashed signatures.  If we follow that guidance, then like the algorithm variant, then pure vs. pre-hashed tells you how you can use the key.

With that in mind, recall the OTP structure that encodes the key:
```
SPX_KEY_TYPEn: a word associating a key with the TEST, DEV or PROD lifecycle states.
SPX_KEYn: 32 bytes of key material
SPX_KEY_CONFIGn: a word identifying the key algorithm (SHAKE-128s, SHA2-128s, etc).
```

I think we can dual-purpose the `CONFIGn` word to encode both the algorithm and key-usage constraint.  That is to say:
```
enum key_config_t {
  kShake_128s_pure,
  kShake_128s_prehashed,
  kSha2_128s_pure,
  kSha2_128s_prehashed,
  ...
}
```

Then plumb the [`config_id` field](https://github.com/lowRISC/opentitan/blob/6e698b4dfec2f9e61e091cbb9a866197681745f9/sw/device/silicon_creator/rom/sigverify_key_types.h#L223-L236) through and use it to call the verify function appropriately.

Comment (moidx) [reply hw/ip/otp_ctrl/data/BUILD:99]: It is almost there... 

https://github.com/lowRISC/opentitan/blob/6e698b4dfec2f9e61e091cbb9a866197681745f9/sw/device/silicon_creator/rom/sigverify_key_types.h#L223-L236

File: sw/device/silicon_creator/lib/sigverify/spx_verify.c
Comment (moidx) [inline sw/device/silicon_creator/lib/sigverify/spx_verify.c:87]: I think this is the right byte ordering. Also the right OID identifier for `SHA-2-256`.

```
OID ← toByte(0x0609608648016503040201, 11) ▷ 2.16.840.1.101.3.4.2.1
```

**Review by moidx — APPROVED** [review]

**Comment by jadephilipoom** [comment]
Moving out of draft since https://github.com/lowRISC/opentitan/pull/23762 is now merged

File: sw/device/silicon_creator/lib/sigverify/spx_verify.c
Comment (cfrantz) [inline sw/device/silicon_creator/lib/sigverify/spx_verify.c:93]: The `config_id_t` is an integer; I don't think we need to pass a pointer to it here.

File: sw/device/silicon_creator/rom/rom.c
Comment (cfrantz) [inline sw/device/silicon_creator/rom/rom.c:333]: Since this value is just an integer, we probably don't need a pointer to it here -- we can just get a copy of it from `spx_key_get` and pass-by-value to `spx_verify`.

If you're concerned with having an initial invalid value, zero or all-ones would be perfectly good as invalid values.

File: sw/host/opentitanlib/src/crypto/spx.rs
Comment (cfrantz) [inline sw/host/opentitanlib/src/crypto/spx.rs:98]: How do you want to connect this up to the CLI?

Obviously, we could have a command-line switch that selects among raw, pure and pre-hash.  We could also attach some meta-data to the key to indicate its usage constraint.

**Comment by jadephilipoom** [comment]
Resolves https://github.com/lowRISC/opentitan/issues/23144

In the TC we agreed to switch to SHA2 parameters and then implement an option that allowed us to select either SHA2 or SHAKE if there was enough time before ROM freeze and enough ROM space. However, I think at this point there is not enough time before ROM freeze -- I could _maybe_ implement the switching in time, but we'd be merging it at the very last second and I think we're better off focusing on testing this version instead. So once this is merged, the SPHINCS+ updates for EarlGrey prod should be complete.

File: sw/host/opentitanlib/src/crypto/spx.rs
Comment (jadephilipoom) [reply sw/host/opentitanlib/src/crypto/spx.rs:98]: Yes, I think a mode option with values raw, pure, and pre-hash, defaulting to pure, would make sense.

File: sw/host/opentitanlib/src/crypto/spx.rs
Comment (cfrantz) [inline sw/host/opentitanlib/src/crypto/spx.rs:41]: I noticed that the `domain_sep` in the verify_prehash is different from the one in the `sign_prehash` function: `[1, 8]` vs `[0, 8]`.  Typo?

Comment (jadephilipoom) [reply sw/host/opentitanlib/src/crypto/spx.rs:41]: Yep, a typo. Fixed and I've now added a unit test here that makes sure `sign_prehash` and `verify_prehash` agree!

**Comment by jadephilipoom** [comment]
I've added a new commit that just increases the timeout to "long" for the HMAC functest in silicon_creator; since I added a few tests, it seems like it's very close to the timeout limit and I saw it time out once here.

**Review by cfrantz — APPROVED** [review]

**Review by moidx — APPROVED** [review]
🎉🎉🎉🎉!

**Pull request merged**
